File: static/js/fiat_price.js
/**
 * Fiat → BTC helpers (mirrors trato/services/fiat_price.py).
 * Loaded before index.js — one sats-from-fiat path for cards, detail, take flow.
 */
;(function (root) {
  const FIAT_PRICE_ALIASES = {
    USDT: 'USD',
    USDC: 'USD',
    'L-USDT': 'USD',
    'L-USDC': 'USD',
    BUSD: 'USD',
    DAI: 'USD',
    TUSD: 'USD',
    USDP: 'USD',
    GUSD: 'USD',
    PYUSD: 'USD'
  }
  const SUSPICIOUS_IMPLICIT = new Set(['ALL'])
  const GLOBAL_FALLBACKS = ['USD', 'EUR']

  function normalizeFiatCode(code) {
    return String(code || 'USD')
      .trim()
      .toUpperCase()
  }

  function resolvePriceFiatCode(fiatCode) {
    const code = normalizeFiatCode(fiatCode)
    return FIAT_PRICE_ALIASES[code] || code
  }

  function isSafeImplicitFallback(code, primary) {
    const c = normalizeFiatCode(code)
    const p = normalizeFiatCode(primary)
    if (SUSPICIOUS_IMPLICIT.has(c) && c !== p) return false
    return Boolean(c)
  }

  function priceFiatFallbackChain(fiatCode, opts) {
    opts = opts || {}
    const primary = normalizeFiatCode(fiatCode)
    const chain = []
    const push = c => {
      const code = normalizeFiatCode(c)
      if (code && chain.indexOf(code) === -1) chain.push(code)
    }
    push(resolvePriceFiatCode(primary))
    if (primary !== resolvePriceFiatCode(primary)) push(primary)
    ;[opts.walletFiat, opts.instanceFiat].forEach(extra => {
      if (!extra || !isSafeImplicitFallback(extra, primary)) return
      push(extra)
      const alias = resolvePriceFiatCode(extra)
      if (alias !== normalizeFiatCode(extra)) push(alias)
    })
    if (opts.includeGlobal !== false) {
      GLOBAL_FALLBACKS.forEach(fb => push(fb))
    }
    if (primary === 'ALL') push('ALL')
    return chain
  }

  function orderPriceFactor(premiumPct) {
    return 1 - (Number(premiumPct) || 0) / 100
  }

  function satsFromFiat(fiatAmount, satsPerFiat, premiumPct) {
    const fiat = Number(fiatAmount)
    const rate = Number(satsPerFiat)
    if (!fiat || fiat <= 0 || !rate || rate <= 0 || isNaN(fiat) || isNaN(rate)) {
      return 0
    }
    return Math.max(
      0,
      Math.round(fiat * rate * orderPriceFactor(premiumPct))
    )
  }

  function formatSats(n) {
    if (n == null || n === '' || Number.isNaN(Number(n))) return ''
    return Math.round(Number(n)).toLocaleString('en-US', {
      maximumFractionDigits: 0,
      minimumFractionDigits: 0,
      useGrouping: true
    })
  }

  function quoteCacheKey(order) {
    if (!order) return ''
    return `${String(order.id || '')}:${normalizeFiatCode(order.fiat_code)}`
  }

  function pickOrderFiatAmount(order, ctx) {
    ctx = ctx || {}
    if (!order) return null
    if (order.is_range) {
      const raw =
        ctx.takeFiatAmount != null ? ctx.takeFiatAmount : order.fiat_min
      const v = Number(raw)
      if (isNaN(v)) return null
      if (ctx.inRange && !ctx.inRange(order, v)) return null
      return v
    }
    if (order.fiat_amount != null && order.fiat_amount !== '') {
      const v = Number(order.fiat_amount)
      return isNaN(v) ? null : v
    }
    return null
  }

  function isPlausibleSatsForFiat(amountSats, fiatAmount, satsPerFiat, premiumPct) {
    const sats = Number(amountSats) || 0
    const fiat = Number(fiatAmount) || 0
    const rate = Number(satsPerFiat) || 0
    if (sats <= 0 || fiat <= 0 || rate <= 0) return true
    const expected = satsFromFiat(fiat, rate, premiumPct)
    if (expected <= 0) return true
    const ratio = sats / expected
    return ratio >= 0.25 && ratio <= 4
  }

  function estimateOrderBtcSats(order, ctx) {
    ctx = ctx || {}
    if (!order) return null
    const fixed = Number(order.amount_sats) || 0
    const market =
      typeof ctx.orderUsesMarketPrice === 'function'
        ? ctx.orderUsesMarketPrice(order)
        : Boolean(order.is_market_price || !(fixed > 0))
    if (fixed > 0 && !market) {
      const fiat = pickOrderFiatAmount(order, ctx)
      const code = normalizeFiatCode(order.fiat_code)
      let rate = null
      if (ctx.quoteSatsPerFiat && ctx.quoteFiatCode === code) {
        rate = Number(ctx.quoteSatsPerFiat)
      } else if (ctx.orderQuotes) {
        const cached = ctx.orderQuotes[quoteCacheKey(order)]
        if (cached && cached.fiat_code === code && cached.sats_per_fiat) {
          rate = Number(cached.sats_per_fiat)
        }
      }
      if (
        fiat &&
        rate &&
        rate > 0 &&
        !isPlausibleSatsForFiat(fixed, fiat, rate, Number(order.premium) || 0)
      ) {
        return satsFromFiat(fiat, rate, Number(order.premium) || 0)
      }
      return fixed
    }

    const fiat = pickOrderFiatAmount(order, ctx)
    const code = normalizeFiatCode(order.fiat_code)
    let rate = null
    if (ctx.quoteSatsPerFiat && ctx.quoteFiatCode === code) {
      rate = Number(ctx.quoteSatsPerFiat)
    } else if (ctx.orderQuotes) {
      const cached = ctx.orderQuotes[quoteCacheKey(order)]
      if (cached && cached.fiat_code === code && cached.sats_per_fiat) {
        rate = Number(cached.sats_per_fiat)
      }
    }
    if (!rate && ctx.marketPrice) {
      const mp = ctx.marketPrice
      if (normalizeFiatCode(mp.fiat_code) === code && mp.sats_per_fiat) {
        rate = Number(mp.sats_per_fiat)
      }
    }
    if (!fiat || !rate || rate <= 0) return fixed > 0 ? fixed : null
    return satsFromFiat(fiat, rate, Number(order.premium) || 0)
  }

  const api = {
    FIAT_PRICE_ALIASES,
    normalizeFiatCode,
    resolvePriceFiatCode,
    priceFiatFallbackChain,
    orderPriceFactor,
    satsFromFiat,
    formatSats,
    quoteCacheKey,
    pickOrderFiatAmount,
    isPlausibleSatsForFiat,
    estimateOrderBtcSats
  }

  root.TratoFiatSats = api
  root.TratoFiatPrice = api
})(typeof window !== 'undefined' ? window : globalThis)

File: services/fiat_price.py
"""Central fiat→BTC price resolution for Trato.

LNbits has no BTC/USDT feed; stablecoins map to USD. When an instance only
allows ``ALL`` (Albanian Lek, not "all currencies"), applying that rate to a
USD/USDT book row yields ~9 sats per dollar instead of ~1000 — fees show as
~540 sats instead of ~100k+ BTC on a 100 USDT trade.
"""

from __future__ import annotations

from typing import Any

# Book / wallet codes → ISO code LNbits can price.
FIAT_PRICE_ALIASES: dict[str, str] = {
    "USDT": "USD",
    "USDC": "USD",
    "L-USDT": "USD",
    "L-USDC": "USD",
    "BUSD": "USD",
    "DAI": "USD",
    "TUSD": "USD",
    "USDP": "USD",
    "GUSD": "USD",
    "PYUSD": "USD",
}

_GLOBAL_FALLBACKS = ("USD", "EUR")


def normalize_fiat_code(fiat_code: str | None) -> str:
    return (fiat_code or "USD").strip().upper()


def resolve_fiat_price_code(fiat_code: str | None) -> str:
    code = normalize_fiat_code(fiat_code)
    if code in FIAT_PRICE_ALIASES:
        return FIAT_PRICE_ALIASES[code]
    if code.replace("-", "") == "LUSDT":
        return "USD"
    return code


resolve_price_fiat_code = resolve_fiat_price_code


def resolve_fiat_code(fiat_code: str) -> tuple[str, str]:
    """Return ``(lookup_code, requested_code)`` for exchange-rate lookups."""
    requested = normalize_fiat_code(fiat_code)
    return resolve_fiat_price_code(requested), requested


def premium_price_factor(premium_pct: float) -> float:
    """+premium % means fewer sats per fiat unit (maker values BTC higher)."""
    return 1.0 - float(premium_pct) / 100.0


def sats_per_fiat_from_btc_price(btc_price_fiat: float) -> float:
    if btc_price_fiat <= 0:
        return 0.0
    return 100_000_000 / btc_price_fiat


def price_fallback_chain(
    fiat_code: str,
    *,
    instance_fiat: str | None = None,
    wallet_fiat: str | None = None,
) -> list[str]:
    """Ordered fiat codes to try when fetching ``sats_per_fiat``."""
    chain: list[str] = []

    def push(c: str | None) -> None:
        c = (c or "").strip().upper()
        if c and c not in chain:
            chain.append(c)

    raw = normalize_fiat_code(fiat_code)
    resolved = resolve_fiat_price_code(raw)
    push(resolved)
    if raw != resolved:
        push(raw)
    for c in (wallet_fiat, instance_fiat):
        c = (c or "").strip().upper()
        if c and c != "ALL":
            push(c)
            alias = resolve_fiat_price_code(c)
            if alias != c:
                push(alias)
    for fb in _GLOBAL_FALLBACKS:
        push(fb)
    if raw == "ALL" or resolved == "ALL":
        push("ALL")
    return chain


price_fiat_fallback_chain = price_fallback_chain


async def fetch_fiat_price_quote(
    fiat_code: str,
    *,
    instance_fiat: str | None = None,
    wallet_fiat: str | None = None,
) -> dict[str, Any]:
    """Resolve aliases and fallbacks; return LNbits rate metadata for the UI."""
    from lnbits.utils.exchange_rates import get_fiat_rate_and_price_satoshis

    requested = normalize_fiat_code(fiat_code)
    resolved = resolve_fiat_price_code(requested)
    chain = price_fallback_chain(
        requested,
        instance_fiat=instance_fiat,
        wallet_fiat=wallet_fiat,
    )
    for try_code in chain:
        if try_code == "ALL" and resolved != "ALL" and requested != "ALL":
            continue
        try:
            rate, price = await get_fiat_rate_and_price_satoshis(try_code)
        except Exception:
            continue
        if not rate or rate <= 0:
            continue
        return {
            "fiat_code": requested,
            "price_fiat_code": try_code,
            "sats_per_fiat": float(rate),
            "btc_price": float(price),
            "resolved_via_alias": try_code == resolved and requested != resolved,
            "fallback_chain": chain,
        }
    raise ValueError(f"No exchange rate for {requested}")


def sats_from_fiat(
    fiat_amount: float,
    sats_per_fiat: float,
    *,
    premium_pct: float = 0.0,
) -> int:
    """Satoshis for a fiat amount at market, with maker premium applied."""
    if fiat_amount <= 0 or sats_per_fiat <= 0:
        return 0
    factor = premium_price_factor(premium_pct)
    return max(0, round(float(fiat_amount) * float(sats_per_fiat) * factor))


fiat_to_sats = sats_from_fiat


def sats_to_fiat(
    amount_sats: int,
    sats_per_fiat: float,
    *,
    premium_pct: float = 0.0,
) -> float:
    if amount_sats <= 0 or sats_per_fiat <= 0:
        return 0.0
    factor = premium_price_factor(premium_pct) or 1.0
    return round((amount_sats / sats_per_fiat / factor) * 100) / 100


def format_sats_display(value: int | float | None) -> str:
    """Locale-neutral integer sats string (comma thousands, never ``540.501``)."""
    if value is None:
        return "—"
    try:
        n = int(round(float(value)))
    except (TypeError, ValueError):
        return "—"
    return f"{n:,}"


def is_plausible_sats_for_fiat(
    amount_sats: int,
    fiat_amount: float,
    sats_per_fiat: float,
    *,
    premium_pct: float = 0.0,
) -> bool:
    """Reject platform ``amount_sats`` that disagree with fiat × rate by >4×."""
    if amount_sats <= 0 or fiat_amount <= 0 or sats_per_fiat <= 0:
        return True
    expected = sats_from_fiat(fiat_amount, sats_per_fiat, premium_pct=premium_pct)
    if expected <= 0:
        return True
    ratio = amount_sats / expected
    return 0.25 <= ratio <= 4.0
